### ISSUE TITLE
Relation between Models breaks when using casted columns via CastsAttributes

### ISSUE THREAD

**Issue by VictorDauchy** [laravel/framework#32021]
- Laravel Version: 7.1.0
- PHP Version: 7.4.3
- Database Driver & Version: MySql 5.7

### Description:

When 2 models are linked by columns using the new casting feature (Using CastsAttributes interface) the relation access breaks.

```
    public function foo(): BelongsTo {
        return $this->belongsTo(
            Foo::class,
            'foreignCastedColumnName',
            'ownerCastedColumnName');
    }
```
The error is: 
```
In BelongsTo.php line 181:

  [ErrorException]
  Illegal offset type


Exception trace:
  at /www/htdocs/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Relations/BelongsTo.php:181
 Illuminate\Foundation\Bootstrap\HandleExceptions->handleError() at /www/htdocs/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Relations/BelongsTo.php:181
 Illuminate\Database\Eloquent\Relations\BelongsTo->match() at /www/htdocs/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Builder.php:586
 Illuminate\Database\Eloquent\Builder->eagerLoadRelation() at /www/htdocs/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Builder.php:556
 Illuminate\Database\Eloquent\Builder->eagerLoadRelations() at /www/htdocs/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Builder.php:524
 Illuminate\Database\Eloquent\Builder->get() at /www/htdocs/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Model.php:486


```

In this case both columns are simple 2 characters strings in database (Representing country ISO code),.
The relation is properly enforced and works at MySql level. 
However the Cast transforms the 2 characters string to a Country object.

The crash is due to the fact that the function ```BelongsTo::match(array $models, Collection $results, $relation)``` uses values after casting.

This seem to solve the issue (using ```getAttributes()``` to access attributes before cast):

```
    public function match(array $models, Collection $results, $relation)
    {
        $foreign = $this->foreignKey;

        $owner = $this->ownerKey;

        // First we will get to build a dictionary of the child models by their primary
        // key of the relationship, then we can easily match the children back onto
        // the parents using that dictionary and the primary key of the children.
        $dictionary = [];

        foreach ($results as $result) {
            $dictionary[$result->getAttributes()[$owner]] = $result;
        }

        // Once we have the dictionary constructed, we can loop through all the parents
        // and match back onto their children using these keys of the dictionary and
        // the primary key of the children to map them onto the correct instances.
        foreach ($models as $model) {
            $foreignAttribut = $model->getAttributes()[$foreign];
            if (isset($dictionary[$foreignAttribut])) {
                $model->setRelation($relation, $dictionary[$foreignAttribut]);
            }
        }

        return $models;
    }
```

**Comment by driesvints** [comment]
Don't have time for this myself atm unfortunately. Appreciating any help with this.

@staudenmeir perhaps?

**Comment by staudenmeir** [comment]
I'll look into it.

**Comment by VictorDauchy** [comment]
Little up as I'm having the same issue with HasManyThrough.
I suppose the issue can be assumed to be present in all:
```
BelongsTo.php
BelongsToMany.php
HasMany.php
HasManyThrough.php
HasOne.php
HasOneThrough.php
MorphMany.php
MorphOne.php
MorphTo.php
```
Do you need some help? I never contributed but if I can do something...

**Comment by staudenmeir** [comment]
Yes, this affects all relations. Ideally, PHP would convert objects with a `__toString()` method to a string when being used as an array offset...

I see three options:
1. Using the raw attribute.
2. Casting the attribute to a string.
3. Only casting the attribute to a string if it's an object.

The last one would be the safest regarding backward compatibility but adds some overhead to a method that is used *a lot*.

The issue with the first option is that [`getAttributes()`](https://github.com/laravel/framework/blob/2b395cd1f2fe95b67edf97684f09b7c5c4a55152/src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php#L1122) is not the simple method it was in previous versions of Laravel.

I think the best solution would be a method like `getRawAttribute()`, but this particular name would be a breaking change (conflicts with [accessors](https://laravel.com/docs/7.x/eloquent-mutators#defining-an-accessor) for a `raw` attribute).

**Comment by VictorDauchy** [comment]
Hi,

As I wrote, the use of the __toString() method may just create other issues because there is no guaranty that it will return the the key matching the database record.

For example:
In database: the key could be `ca` and is used as foreign key.
The cast could create a country object like: `{iso: 'ca'}`
And the __toString() may return: `strtoupper($this->iso)`

So this would crash again. Plus why force people to create objects which implement __toString(), stringable, etc... ? That sound like a hidden expectation. So to my mind, options 2 and 3 would be out.

To me the only way is to use a value matching the database record, so a raw value from attributes array.

I agree with your conclusion about having a `getRawAttribute`, to me that would be the best solution but again as you said that would conflict with the getXXXAttribute magic.

What is your issue with `getAttributes` as a fix for now ? Then maybe a better solution for Laravel 8 ?

**Comment by VictorDauchy** [comment]
To keep on your hybrid solution (3) but applied to option (1) AKA:

Use `getAttributes()[$key]` only if it is an object.

To my mind that does not guaranty the safety of it as a caster like:

```
{
    /**
     * @inheritDoc
     */
    public function get($model, string $key, $value, array $attributes): string {
        return Str::snake($value);
    }

    /**
     * @inheritDoc
     */
    public function set($model, string $key, $value, array $attributes): string {
        return Str::camel($value);
    }
}
```
Would fail again as the string used in the relation would be `snake` case while in database the foreign key would apply on the `camel` case string.
Sure the error would not be the same but the relation would 'fail' silently.

**Comment by VictorDauchy** [comment]
An other thought would be to not limit ourself to string type. 
For example a key could be the the numeric code of a currency (https://en.wikipedia.org/wiki/ISO_4217) being cast as a Currency object.

**Comment by staudenmeir** [comment]
Actually, we can't use the raw attribute: I've seen cases where people rely on accessors being called (e.g. to implement JSON relationships).

So we could solve cases where the cast returns a `Stringable` instance (but doesn't change the value) with option 2 or 3.

> however it may return a transformed string, different than what is used in database.

Maybe that's an edge case we just can't support. Personally, I would advise against implementing country/currency/language foreign keys with strings (but that's a whole other discussion).

**Comment by VictorDauchy** [comment]
I feel that to make the best choice and avoid long discussion other contributors might give there thoughts... Like @driesvints @taylorotwell ...
I don't know you manage that kind of situation...

**Comment by VictorDauchy** [comment]
Small update: On my side to handle the potential performance issue (and since we cannot create a function `getRawAttributes()`) I created:

```
    /**
     * Get a subset of the model's raw attributes.
     *
     * @param  array|mixed  $attributes
     * @return array
     */
    public function onlyRaw($attributes)
    {
        $attributes = is_array($attributes) ? $attributes : func_get_args();

        $this->mergeAttributesFromClassCasts($attributes);

        return array_intersect_key($this->attributes, array_flip($attributes));
    }
```

Which mimic the existing `HasAttributes::only($attributes)` method.

To handle that I added an optional `$filter` parameter to `HasAttributes::mergeAttributesFromClassCasts` method as:

```
    /**
     * Merge the cast class attributes back into the model.
     *
     * @param array|null $filter
     * @return void
     */
    protected function mergeAttributesFromClassCasts(?array $filter = null)
    {
        $filteredCast = $filter ? array_intersect_key($this->classCastCache, array_flip($filter)) : $this->classCastCache;
        foreach ($filteredCast as $key => $value) {
            $caster = $this->resolveCasterClass($key);

            $this->attributes = array_merge(
                $this->attributes,
                $caster instanceof CastsInboundAttributes
                    ? [$key => $value]
                    : $this->normalizeCastClassResponse($key, $caster->set($this, $key, $value, $this->attributes))
            );
        }
    }
```

So now we only cast necessary attributes and have the right value doing:

```
$model->onlyRaw($foreign)[$foreign];
```

Any thoughts on this implementation @staudenmeir ?
